### PR TITLE
re:  Simplify `PeerActorManager` by removing `IncomingCrutch`

### PR DESCRIPTION
**Pull request by pmnoxx**
We can remove `IncomingCrutch` and simplify `PeerManagerActor`. There is no reason, why we should be using `IncomingCrutch`. We can get rid of one atomic, at the same time.

Based on https://github.com/near/nearcore/pull/5645/files

### CODE REVIEW

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 1]
```diff
@@ -2268,7 +2240,11 @@ impl PeerManagerActor {
                 PeerManagerMessageResponse::OutboundTcpConnect(())
             }
             PeerManagerMessageRequest::InboundTcpConnect(msg) => {
-                self.handle_msg_inbound_tcp_connect(msg, ctx);
+                if self.peer_counter.load(Ordering::SeqCst)
+                    < self.config.max_num_peers as usize + LIMIT_PENDING_PEERS
+                {
+                    self.handle_msg_inbound_tcp_connect(msg, ctx);
+                }
```
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:2247]: so here we'd silently drop the connection if the limit is exceeded? might make debugging a lot harder. 

We should add some log here (and make sure that it is not printed too often)

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 2]
```diff
@@ -1503,38 +1482,32 @@ impl Actor for PeerManagerActor {
...
+                } else {
+                    return;
+                };
+
+                loop {
+                    let (conn, client_addr) =
+                        if let Ok((conn, client_addr)) = listener.accept().await {
+                            (conn, client_addr)
+                        } else {
+                            return;
```
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:1501]: so any error on the listener part will cause us to stop listening at all?

or is there some mechanism that calls 'started' again?

**Review by mm-near — CHANGES_REQUESTED** [review]

File: chain/network/src/peer_manager/peer_manager_actor.rs (hunk 1 above)
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:2297]: Yes, that's how the current logic works. We can discuss changing it in another PR.

File: chain/network/src/peer_manager/peer_manager_actor.rs (hunk 2 above)
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs]: Currently, we continue on error. I changed the logic to be the same as before.

We will continue listening, and ignore all failures of `listener.accept()`. Same as before.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 3]
```diff
@@ -1503,38 +1482,32 @@ impl Actor for PeerManagerActor {
...
-                    }));
-                    actix::fut::ready(())
-                },
-            ));
+            async move {
+                let listener = TcpListener::bind(server_addr).await.ok();
+                let listener = if let Some(listener) = listener {
+                    listener
+                } else {
+                    return;
```
[the remark this answers is not in the document]
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs]: @mm-near I changed the logic to be the same as before. We will panic in case of failure.

**Comment by pmnoxx** [comment]
@matklad Can you take a look? This part of code is much simpler than before, but it's good if you take a look at is as well.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 4]
```diff
@@ -800,10 +795,10 @@ impl PeerManagerActor {
...
                     .take_while(|x| match x {
-                        Ok(_) => future::ready(true),
+                        Ok(_) => true,
                         Err(e) => {
                             warn!(target: "network", ?e, "Peer stream error");
-                            future::ready(false)
+                            false
                         }
                     })
                     .map(Result::unwrap),
```
Comment (matklad) [inline chain/network/src/peer_manager/peer_manager_actor.rs:804]: One of those rare cases where there *should* be a better way to do this, but there isn't one currently. Opened https://github.com/tokio-rs/tokio/issues/4337.

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 5]
```diff
@@ -1508,62 +1503,38 @@ impl PeerManagerActor {
...
-                },
-            ));
+                loop {
+                    if let Ok((conn, client_addr)) = listener.accept().await {
+                        peer_manager_addr.do_send(PeerManagerMessageRequest::InboundTcpConnect(
+                            InboundTcpConnect::new(conn),
+                        ));
+                        debug!(target: "network", from = ?client_addr, "got new connection");
+                    }
+                }
```
Comment (matklad) [inline chain/network/src/peer_manager/peer_manager_actor.rs:1534]: We don't seem to ever break out of this loop? Ie, when `accept()` returns an error, we will start busy-looping here. It's also not super clear if this handles cancellation for the happy case? How one does actually stop PeerManager?

Comment (matklad) [reply chain/network/src/peer_manager/peer_manager_actor.rs:1534]: Though it seems like pre-existing logic, so probably is out of scope for the PR.

Comment (matklad) [reply chain/network/src/peer_manager/peer_manager_actor.rs:1534]: Ok, upon further investigation, it is probably fine -- at the top-level, when we catch `ctlr_c`, we ask the System to stop, and presumably it should stop polling the future.

**Review by matklad — APPROVED** [review]

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near: /chain/network/
```
@bowenwang1996 @mm-near Can you take a look?

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near: chain/network/
```
@bowenwang1996 @mm-near Can you take a look?

File: chain/network/src/peer_manager/peer_manager_actor.rs (hunk 4 above)
Comment (matklad) [reply chain/network/src/peer_manager/peer_manager_actor.rs:878]: btw, this was fixed upstream

Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:875]: @matklad Thanks. We can make it even better in the future.

**Review by bowenwang1996 — COMMENTED** [review]
Thanks!

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**